storage/src/main/java/io/delta/storage/internal/ThreadUtils.java

```diff
@@ -0,0 +1,113 @@
+        // since values used in static method must be final
+        List<Throwable> exceptionHolder = new ArrayList<>(1);
+        List<T> resultHolder = new ArrayList<>(1);
+        Thread thread = new Thread(threadName) {
+            @Override
+            public void run() {
+                try {
+                    resultHolder.add(body.call());
+                } catch (Throwable t) {
+                    if (LogStoreErrors.isNonFatal(t)) {
+                        exceptionHolder.add(t);
+                    }
+                }
+            }
+        };
+        thread.setDaemon(isDaemon);
+        thread.start();
+        try {
+            thread.join();
+        } catch (InterruptedException e) {
+            throw new InterruptedIOException(e.getMessage());
+        }
+        if (!exceptionHolder.isEmpty()) {
+            Throwable realException = exceptionHolder.get(0);
+            // Remove the part of the stack that shows method calls into this helper method
+            // This means drop everything from the top until the stack element
+            // ThreadUtils.runInNewThread(), and then drop that as well (hence the + 1 to start index).
+            List<StackTraceElement> currentThreadStackTrace =
+                Arrays.asList(Thread.currentThread().getStackTrace());
+            // index of the stack element pointing to ThreadUtils.runInNewThread()
+            int startIndex = currentThreadStackTrace.stream()
+                .filter(e -> e.getClassName().contains(ThreadUtils.class.getSimpleName()))
+                .map(currentThreadStackTrace::indexOf)
+                .findFirst()
+                .orElse(-1);
+            List<StackTraceElement> baseStackTrace;
+            if (startIndex == -1 || startIndex >= currentThreadStackTrace.size() - 1) {
+                baseStackTrace = Collections.emptyList();
+            } else {
+                // startIndex + 1 to drop the stack element ThreadUtils.runInNewThread()
+                baseStackTrace =
+                    currentThreadStackTrace.subList(startIndex + 1, currentThreadStackTrace.size());
+            }
+            // Remove the part of the new thread stack that shows methods call from this helper method
+            // This means take everything from the top until the stack element
+            StackTraceElement[] realExceptionStackTrace = realException.getStackTrace();
+            List<StackTraceElement> extraStackTrace = new ArrayList<>();
+            for (StackTraceElement st : realExceptionStackTrace) {
+                if (!st.getClassName().contains(ThreadUtils.class.getSimpleName())) {
+                    extraStackTrace.add(st);
+                }
+            }
+            // Combine the two stack traces, with a place holder just specifying that there
+            // was a helper method used, without any further details of the helper
+            List<StackTraceElement> placeHolderStackElem = Arrays.asList(
```

Comment:
nit: we don't need to create an array here. We can just call `finalStackTrace.add(placeHolderStackElem)` below if `placeHolderStackElem` is a `StackTraceElement`.